src/bgw/job_stat.c

```diff
@@ -161,11 +161,21 @@ typedef struct
 static TimestampTz
 calculate_next_start_on_success(TimestampTz last_finish, BgwJob *job)
 {
-	/* TODO: add randomness here? Do we need a range or just a percent? */
-	TimestampTz ts =
-		DatumGetTimestampTz(DirectFunctionCall2(timestamptz_pl_interval,
-												TimestampTzGetDatum(last_finish),
-												IntervalPGetDatum(&job->fd.schedule_interval)));
+	TimestampTz ts;
+	PG_TRY();
+	{
+		ts =
+			DatumGetTimestampTz(DirectFunctionCall2(timestamptz_pl_interval,
+													TimestampTzGetDatum(last_finish),
+													IntervalPGetDatum(&job->fd.schedule_interval)));
+	}
+	PG_CATCH();
+	{
+		FlushErrorState();
+		ts = ts_timer_get_current_timestamp();
```

Comment:
I'm wondering if the current timestamp is the correct value to set the next start to if `timestamptz_pl_interval` fails. From what I've seen in the code, this can happen in two ways:

1.  `timestamp2tm` fails on a negative Julian days, which shouldn't (?) happen
2. `timestamptz_pl_interval` generates a timestamp that's too large or small to be valid.

There's no good reason for `schedule_interval` to be negative, 2 should only be due to overflow, in which case it seems like the natural time to reschedule is the max timestamp. If we use the current timestamp, the job will immediately run again, and if it succeeds, will likely run into this same case, and be continually be scheduled.

Reply:
@JLockerman What is max timestamp?

Reply:
AIUI, if the next timestamp is set to max timestamp, then it will never be rescheduled. This is an error condition, so please print an error message in that case.